project/src/js/ui/tournament/detail/tournamentDetailCtrl.js

```diff
@@ -24,33 +26,66 @@ export default function controller() {
     if (data.secondsToStart && data.secondsToStart > 0) {
       data.secondsToStart--;
     }
-
     if (data.secondsToFinish && data.secondsToFinish > 0) {
       data.secondsToFinish--;
     }
-
     m.redraw();
   }
 
+  function join(id) {
+    if (!id) {
+      id = tournament().id;
+    }
+    xhr.join(id).then(() => {
+      hasJoined(true);
+      m.redraw();
+    }, err => utils.handleXhrError(err));
+  }
+
+  function withdraw(id) {
+    if (!id) {
+      id = tournament().id;
+    }
+    xhr.withdraw(id).then(() => {
+      hasJoined(false);
+      m.redraw();
+    }, err => utils.handleXhrError(err));
+  }
+
+  let id = m.route.param('id');
+  let action = m.route.param('action');
+  if (action && action === 'withdraw') {
```

Comment:
I'm not sure having an `action` route parameter is the right way to do this. There is no such other behavior in the app. This is a client side app and url params should only have ids or params used to read data, not perform actions (contrary to what you would do server side).

The withdraw action is actually carried by the server. So your job is to send the xhr and update your business data after server response. Which is already what you do, so you just need to do it directly without the mithril route (this is not a good practice to do so in mithril I believe).